packages/react/toggle-button/src/ToggleButton.tsx

```diff
@@ -1,10 +1,199 @@
+      }
+
+      setValue(
+        value.includes(buttonValue)
+          ? value.filter((v) => v !== buttonValue)
+          : value.concat(buttonValue).sort()
+      );
+    }
+
+    const context: ToggleButtonGroupContextValue = {
+      setValue,
+      value,
+      handleChange,
+      rovingFocus,
+      name: GROUP_CONTEXT_NAME,
+    };
+
+    return (
+      <Comp {...getPartDataAttrObj(GROUP_NAME)} role="group" ref={forwardedRef} {...groupProps}>
+        <ToggleButtonGroupContext.Provider value={context}>
+          <ToggleButtonGroupInner rovingFocus={rovingFocus}>{children}</ToggleButtonGroupInner>
+        </ToggleButtonGroupContext.Provider>
+      </Comp>
+    );
+  }
+);
+
+function ToggleButtonGroupInner(props: { rovingFocus: boolean; children: React.ReactNode }) {
+  return props.rovingFocus ? (
+    <RovingFocusGroup loop>{props.children}</RovingFocusGroup>
+  ) : (
+    <React.Fragment>{props.children}</React.Fragment>
+  );
+}
+
+/* -------------------------------------------------------------------------------------------------
+ * ToggleButtonExclusive
+ * -----------------------------------------------------------------------------------------------*/
+
+const GROUP_EXC_NAME = 'ToggleButtonGroupExclusive';
+const GROUP_EXC_CONTEXT_NAME = 'ToggleButtonGroupExclusiveContext';
+
+type ToggleButtonGroupExclusiveOwnProps = ToggleButtonGroupSharedProps & {
+  /** The controlled value of the toggled button */
+  value?: string | null;
+  /** The uncontrolled value of the toggled button */
+  defaultValue?: string;
+  /** A function called when the value of the toggled buttons changes */
+  onValueChange?(value: string | null): void;
+};
+
+type ToggleButtonGroupExclusiveContextValue = {
+  value: string | null;
+  setValue:
+    | React.Dispatch<React.SetStateAction<string | null>>
+    | React.Dispatch<React.SetStateAction<string>>;
+  handleChange(value: string | null): void;
+  name: typeof GROUP_EXC_CONTEXT_NAME;
+  rovingFocus: boolean;
+};
+const ToggleButtonGroupExclusiveContext = React.createContext<ToggleButtonGroupExclusiveContextValue | null>(
+  null
+);
+ToggleButtonGroupExclusiveContext.displayName = GROUP_EXC_CONTEXT_NAME;
+
+const ToggleButtonGroupExclusive = forwardRefWithAs<
+  typeof GROUP_DEFAULT_TAG,
+  ToggleButtonGroupExclusiveOwnProps
+>((props, forwardedRef) => {
+  const {
+    as: Comp = GROUP_DEFAULT_TAG,
+    value: valueProp,
+    defaultValue,
+    onValueChange,
+    children,
+    selectionIsRequired = false,
+    rovingFocus = false,
+    ...groupProps
+  } = props;
+
+  const [value, setValue] = useControlledState<string | null>({
+    prop: valueProp,
+    defaultProp: defaultValue || null,
+    onChange: onValueChange,
+  });
+
+  function handleChange(buttonValue: string) {
+    if (!value) {
+      return setValue(buttonValue);
+    }
+
+    if (selectionIsRequired && value === buttonValue) {
+      return;
+    }
+
+    setValue(value === buttonValue ? null : buttonValue);
+  }
```

Comment:
I know this sounds ridiculous but I struggled to follow this block. I was getting mixed up with `value` and `buttonValue` and thought "shouldn't the `required` check be before the `if (!value)` check here?" but I think I understand what's going on after spending some time picking it apart mentally. 

I'm wondering if we could add a var to self document and help folk understand this a bit quicker (and also compute from prev value because of async state again):

```suggestion
  function handleChange(buttonValue: string) {
    setValue(prevValue => {
    	const isRemovingValue = buttonValue && buttonValue === prevValue;
  		return isRemovingValue && !required ? null : buttonValue
  	});
  }
```

Does that cover all the cases you're trying to cover here?